PivotGrid: Add Enhance Performance How To topic (#7546)
DevExpress/devextreme-documentation: concepts/05 UI Components/PivotGrid/155 Enhance Performance on Large Datasets.md

This help topic describes how to improve PivotGrid performance when binding and processing large data sets.

> [note] To enhance component performance, we recommend you process data before binding it to PivotGrid.

Reviewer: what do you mean here? sounds quite abstract. How should I process data if I have performance issues?

Author: Yeah the idea I have in mind for this note is quite abstract. I meant that perhaps users could group/filter their data and create separate, grouped/filtered data stores to use with the component. However, this seems a bit too specific for this tutorial.

### Enable Remote Operations

> The DevExtreme PivotGrid performs all data operations such as filtering and grouping on the client side. To enhance component performance, enable **PivotGridDataSource**.[remoteOperations](/Documentation/ApiReference/Data_Layer/PivotGridDataSource/Configuration/#remoteOperations) to process data on the server side. PivotGrid **remoteOperations** support [ODataStore](/Documentation/ApiReference/Data_Layer/ODataStore/) and [XmlaStore](/Documentation/ApiReference/Data_Layer/XmlaStore/) store types. 

Reviewer: that's not true.

PivotGrid also supports CustomStore (link to API Refs) and createStore (link to https://js.devexpress.com/Angular/Documentation/Guide/Data_Binding/Specify_a_Data_Source/Web_API,_PHP,_MongoDB/). The latter one is quite popular, let's mention this along with a link to a demo: https://js.devexpress.com/Angular/Demos/WidgetsGallery/Demo/PivotGrid/WebAPIService/MaterialBlueLight/

let's remove ODataStore. While it technically works it doesn't support remote grouping and summaries . So PivotGrid will simply load all data from it.

### Enable Virtual Scrolling

Enable virtual scrolling to render cells as users scroll them into view. Set **scrolling**.[mode](/Documentation/ApiReference/UI_Components/dxPivotGrid/Configuration/scrolling/#mode) to *"virtual"*.

### Enable Pagination

> Enable pagination to load data in portions. To do this, enable **PivotGridDataSource**.[paginate](/Documentation/ApiReference/Data_Layer/PivotGridDataSource/Configuration/#paginate).

Reviewer: this works only with OLAP/XmlaStore

This help topic describes how to improve PivotGrid performance when binding and processing large data sets.

### Enable Remote Operations

> The DevExtreme PivotGrid performs all data operations such as filtering and grouping on the client side. To enhance component performance, enable **PivotGridDataSource**.[remoteOperations](/Documentation/ApiReference/Data_Layer/PivotGridDataSource/Configuration/#remoteOperations) to process data on the server side. PivotGrid **remoteOperations** support [XmlaStore](/Documentation/ApiReference/Data_Layer/XmlaStore/) and [CustomStore](/Documentation/ApiReference/Data_Layer/CustomStore/) store types, along with **CustomStore** instances created with [DevExtreme.AspNet.Data](https://github.com/DevExpress/DevExtreme.AspNet.Data/blob/master/README.md).[createStore](https://github.com/DevExpress/DevExtreme.AspNet.Data/blob/master/docs/client-side-with-jquery.md#api-reference).

Reviewer 2: ```suggestion
The DevExtreme PivotGrid performs all data operations, such as filtering and grouping, on the client side. To enhance component performance, enable **PivotGridDataSource**.[remoteOperations](/Documentation/ApiReference/Data_Layer/PivotGridDataSource/Configuration/#remoteOperations) to process data on the server side. PivotGrid **remoteOperations** support [XmlaStore](/Documentation/ApiReference/Data_Layer/XmlaStore/) and [CustomStore](/Documentation/ApiReference/Data_Layer/CustomStore/) store types, along with **CustomStore** instances created with [DevExtreme.AspNet.Data](https://github.com/DevExpress/DevExtreme.AspNet.Data/blob/master/README.md).[createStore](https://github.com/DevExpress/DevExtreme.AspNet.Data/blob/master/docs/client-side-with-jquery.md#api-reference).
```

> To learn more about the **createStore** method, refer to the following topic: [Web API, PHP and MongoDB Data Sources](/Documentation/Guide/Data_Binding/Specify_a_Data_Source/Web_API,_PHP,_MongoDB/)

Reviewer 2: ```suggestion
For more information about the **createStore** method, refer to the following topic: [Web API, PHP and MongoDB Data Sources](/Documentation/Guide/Data_Binding/Specify_a_Data_Source/Web_API,_PHP,_MongoDB/)
```

For more information about integrating a Web API service with PivotGrid, refer to the following demo:

#include btn-open-demo with {
    href: "https://js.devexpress.com/Demos/WidgetsGallery/Demo/PivotGrid/WebAPIService/"
}

### Enable Virtual Scrolling

Enable virtual scrolling to render cells as users scroll them into view. Set **scrolling**.[mode](/Documentation/ApiReference/UI_Components/dxPivotGrid/Configuration/scrolling/#mode) to *"virtual"*.

### Enable Pagination

When binding the component to an OLAP Cube with an **XmlaStore**, you can enable pagination to load data in portions. To do this, enable **PivotGridDataSource**.[paginate](/Documentation/ApiReference/Data_Layer/PivotGridDataSource/Configuration/#paginate).

### Disable the Expand All Button

> The "Expand All" button can request a lot of data from your data source at once. This can strain your server resources and cause other performance issues with your application if you are using a remote data store. Disable [allowExpandAll](/Documentation/ApiReference/UI_Components/dxPivotGrid/Configuration/#allowExpandAll) if you are using large remote datasets to avoid issues with data loading.

Reviewer 2: ```suggestion
The "Expand All" button can request a lot of data from your data source at once. This can strain server resources and cause other performance issues with your application if you are using a remote data store. Disable [allowExpandAll](/Documentation/ApiReference/UI_Components/dxPivotGrid/Configuration/#allowExpandAll) if you are using large remote datasets to avoid issues with data loading.
```
